astr303a.py

```diff
@@ -0,0 +1,571 @@
+import numpy as np
+import pylab as py
+from scipy import optimize
+from matplotlib.patches import Ellipse
+import subprocess
+import sys
+import re
+import os
+import io
+import scipy
+import shutil
+from matplotlib import pylab
+from matplotlib import rc
+from matplotlib import pyplot as plt
+from matplotlib.backends.backend_pdf import PdfPages
+
+#Create a pdf file that all figures will be saved to.
+global pp
+pp = PdfPages('GALAXIES.pdf')
+
+#filename of text file containing SDSS-DR13 data.
+filename = 'specobjdataset.txt'
+
+#read data into memory and organize into an array
+def readData(filename):
+    #read data set into memory
+    with open(filename,'r') as file:
+        data = file.readlines()
+        file.close()
+
+    #parse data into array of galaxies with rows:
+    #[objID(0),ra(1),dec(2),mag(3),magErr(4),ext(5),z(6),zErr(7)]
+    dataarray = []
+    #cycle through all lines of text in specobjdataset.txt
+    for i in range(1,len(data)):
+        line = data[i].replace('\n','') #remove white space and empty lines
+        line = line.split(',') #split lines where commas are found
+        objID = line[6] 
+        ra = float(line[7])
+        dec = float(line[2])
+        mag = float(line[4])
+        magErr = float(line[5])
+        ext = float(line[3])
+        z = float(line [0])
+        zErr = float(line[1])
+        dataarray.append([objID,ra,dec,mag,magErr,ext,z,zErr]) #append galaxy values to array 'dataarray'
+        
+    print('Successfully compiled data!')
+    print(str(len(dataarray))+' galaxies in dataset.')
+    return dataarray
+
+#plot positions of all galaxies
+def part2(dataarray):
+    plt.figure(figsize=(8,8))
+    plt.title('Figure 1: All Galaxies',fontsize=14)
+    plt.ylabel('Declination $(degrees)$',fontsize=12)
+    plt.xlabel('Right Ascension $(degrees)$',fontsize=12)
+    plt.xlim(193.5,196.5)
+    plt.ylim(26.5,29.5)
+    plt.gca().invert_xaxis()
+    
+    #plot individual coordinates
+    for i in range(0,len(dataarray)):
```

Comment:
This line (and many others) shows me that the author doesn't understand fundamentals of Python iteration. One doesn't need to iterate over a range of integers. You can (and should) simply iterate over the values in the list. E.g.:

```python
for _, ra, dec, ... in dataarray:
    plt.plot(ra, dec, color='black', marker='.')
```
Then the code becomes substantially more self-describing. The other option, is if you had used the `CSVDictreader` your data would be self describing and you can simply say:
```python
for item in dataarray:
    plt.plot(item['ra'], item['dec', ...)
```